Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,264 @@
+struct NoDelegate: NIOAsyncSequenceProducerDelegate {
+    func produceMore() {}
+    func didTerminate() {}
+}
+
+/// `AsyncSequence` implementation for handling messages acknowledged by the Kafka cluster (``KafkaAcknowledgedMessage``).
+public struct AcknowledgedMessagesAsyncSequence: AsyncSequence {
+    public typealias Element = Result<KafkaAcknowledgedMessage, KafkaAcknowledgedMessageError>
+    typealias WrappedSequence = NIOAsyncSequenceProducer<Element, NoBackPressure, NoDelegate>
+    let wrappedSequence: WrappedSequence
+
+    /// `AsynceIteratorProtocol` implementation for handling messages acknowledged by the Kafka cluster (``KafkaAcknowledgedMessage``).
+    public struct AcknowledgedMessagesAsyncIterator: AsyncIteratorProtocol {
+        let wrappedIterator: NIOAsyncSequenceProducer<Element, NoBackPressure, NoDelegate>.AsyncIterator
+
+        public mutating func next() async -> Element? {
+            await self.wrappedIterator.next()
+        }
+    }
+
+    public func makeAsyncIterator() -> AcknowledgedMessagesAsyncIterator {
+        return AcknowledgedMessagesAsyncIterator(wrappedIterator: self.wrappedSequence.makeAsyncIterator())
+    }
+}
 
 /// Send messages to the Kafka cluster.
+/// Please make sure to explicitly call ``shutdownGracefully()`` when the `KafkaProducer` is not used anymore.
 /// - Note: When messages get published to a non-existent topic, a new topic is created using the ``KafkaTopicConfig``
 /// configuration object (only works if server has `auto.create.topics.enable` property set).
-public struct KafkaProducer {
-    private let client: KafkaClient
+public actor KafkaProducer {
+    /// States that the ``KafkaProducer`` can have.
+    private enum State {
+        /// The ``KafkaProducer`` has started and is ready to use.
+        case started
+        /// ``KafkaProducer/shutdownGracefully()`` has been invoked and the ``KafkaProducer``
+        /// is in the process of receiving all outstanding acknowlegements and shutting down.
+        case shuttingDown
+        /// The ``KafkaProducer`` has been shut down and cannot be used anymore.
+        case shutDown
+    }
+
+    /// State of the ``KafkaProducer``.
+    private var state: State
+
+    /// Counter that is used to assign each message a unique ID.
+    /// Every time a new message is sent to the Kafka cluster, the counter is increased by one.
+    private var messageIDCounter: UInt = 0
+    /// The ``KafkaTopicConfig`` used for newly created topics.
     private let topicConfig: KafkaTopicConfig
+    /// A logger.
+    private let logger: Logger
+    /// Dictionary containing all topic names with their respective `rd_kafka_topic_t` pointer.
+    private var topicHandles: [String: OpaquePointer]
+
+    // We use implicitly unwrapped optionals here as these properties need to access self upon initialization
+    /// The configuration object of the producer client.
+    private var config: KafkaConfig!
```

Comment:
I think we can get rid of this. The config is already retained in the `KafkaClient` and we never use it outside the init here